Create post Shake Before Use

danielspofford opened the pull request:
It is meant to discuss just what the title suggests, but maybe said a different way: the point in this particular blog post _is not_ to dictate a particular provider. Rather, to point out what is especially important when choosing a provider, given that you will be writing an Elixir service. It is still taking shape so it might look very different from this. There are a handful of blog posts I want to write around this topic so I am struggling in choosing what to say or not in this post.

Yes there are a number of typos, please refrain from calling those out. Rather, content suggestions are highly desired.

In particular, I am curious what else could be included in this post, which areas need buffing up or trimming down, and what seems out of place.

squaresurf commented on content/post/shake-before-use-what-to-consider-when-hosting-elixir.md:
```diff
@@ -0,0 +1,150 @@
+that it will typically be more expensive to scale out on when compared to
+something like AWS or Google Cloud. This concern will become more and more
+important as big cloud providers roll out and buff up their managed Kubernetes
+solutions.
```
These last two sentences could be reworked a bit. It feels like you're mentioning that Heroku handles many details for a developer, but don't really make it clear why it's more expensive to scale out. It would also be good to note why the concern will become more important with the implementation of Kube as a Service.

danielspofford replied:
Yeah not sure yet how I want to rework that paragraph but it needs it.

squaresurf commented on content/post/shake-before-use-what-to-consider-when-hosting-elixir.md:
```diff
@@ -0,0 +1,150 @@
+version via hot code reloading is possible in the world of containers, but it is
+not recommended as it cancels out the idea of immutability and adds in a whole
+host of complexity with respect to knowing what is deployed where, and ensuring
+unexpected restarts bring up the correct version of code.
```
Well said, I love the points and how made them in this paragraph.

danielspofford replied:
👏

danielspofford commented on content/post/shake-before-use-what-to-consider-when-hosting-elixir.md:
```diff
@@ -0,0 +1,150 @@
+If a service is indeed exceptional in that it is imperative that it has
+literally zero downtime and it is stateful like, for example, a telephony
+service, then hot code reloading may be utilized. Typically this lands teams on
+AWS EC2 or other big cloud providers variations thereof. Basically what is
```
providers'

danielspofford commented on content/post/shake-before-use-what-to-consider-when-hosting-elixir.md:
```diff
@@ -0,0 +1,150 @@
+a heroku-esque provider is used: clients of that service are going to have a bad
+time. If the issue is one of latency rather than avoiding a dependency, Redis is
+a good choice here. If that does not fit the bill then an in-memory read-through
+cache backed by persistent storage may be in order. At some point it becomes
```
Such as Redis or Postgres. Reworking a sentence or two here for clarity.

danielspofford commented on content/post/shake-before-use-what-to-consider-when-hosting-elixir.md:
```diff
@@ -0,0 +1,150 @@
+a good choice here. If that does not fit the bill then an in-memory read-through
+cache backed by persistent storage may be in order. At some point it becomes
+more agreeable to simply use a different provider with stronger uptime
+guarantees and potentially a whole host of other features. Something like AWS or
```
The point in the "State in Memory" sections is that holding state in memory (particularly in-application, not something like Redis) is powerful, but is greatly impacted by downtime, be it from bug-induced crashes or daily Heroku restarts.

Furthermore, WRT choosing a provider these limitations don't preclude contenders like Heroku depending on your use case. Then, even if your use case isn't a great fit for Heroku because your state is important and hard to replace, you can circumvent the issue with something like Redis.

This will inevitably reach a point where the pros no longer outweigh the cons and an application will be better served on a different provider that _does_ have strong uptime guarantees.

squaresurf replied:
That makes much more sense. I wonder if you could make some of that explicit in your post. Please forgive me if you already did and I missed it for some reason.

samcdavid commented on content/post/shake-before-use-what-to-consider-when-hosting-elixir.md:
```diff
@@ -0,0 +1,150 @@
+  - Unimportant
+  - Cheap and fast to rebuild
+
+Depending the particular use case and the answer to those questions, it may be
```
Did you mean "Depending on the" 🤔

samcdavid commented on content/post/shake-before-use-what-to-consider-when-hosting-elixir.md:
```diff
@@ -0,0 +1,150 @@
+### Clustering
+
+While it is true that Elixir facilitates scalable, distributed services, it is
+also true that there ain't no such thing as a free lunch. Elixir on a box will
```
While I say `ain't` on a daily basis, we may want to put this in quotes or replace `ain't` with `is`

danielspofford replied:
Your are correct I should quote the quote 😂

samcdavid commented on content/post/shake-before-use-what-to-consider-when-hosting-elixir.md:
```diff
@@ -0,0 +1,150 @@
+It is common to hear about developing stateless APIs in Elixir with Phoenix, in
+that case most providers will suffice accounting for traffic expectations and
+cost of scaling. If a service has more going on than that, it will be more than
+worth it to take the time necesarry to evaluate expected functionality and the
```
`s/necesarry/necessary`

samcdavid commented:
@danielspofford this is really good. I do agree with @squaresurf on that one paragraph that needs some love, but don't really have a better way to state it.

danielspofford commented:
@samcdavid @squaresurf I pushed up a new version that addresses your concerns and clarifies / buffs up some areas. Please review again and comment or approve.
